### ISSUE TITLE
Add async callback support to operator "reduce"

### ISSUE THREAD

**Issue by vitaly-t** [vitaly-t/iter-ops#181]
Most of operators that provide a callback support async result from it, where it may be useful.

There maybe cases when `reduce` could benefit from handling async callbacks also.

**Comment by vitaly-t** [comment]
Closing, for now, because it creates questions like - should the initial value support a promise then also?

It does create some ambiguities, hence the reason why it wasn't supported from start.

**Comment by vitaly-t** [comment]
@RebeccaStevens This one is actually a good exercise for the brain :wink:

So I am re-opening it.

**Comment by vitaly-t** [comment]
b.t.w. what prompted me to get back to this was [this recent question](https://stackoverflow.com/questions/74347704/how-to-return-return-promise-from-array-reduce).

**Comment by vitaly-t** [comment]
I think that in terms of functionality, all it takes is to replace [this line](https://github.com/vitaly-t/iter-ops/blob/main/src/ops/reduce.ts#L114) in async version:

```ts
return this.next();
```

with this one:

```ts
return isPromiseLike(value) ? value.then(() => this.next()) : this.next();
```

And that's it.

However, in terms of function signature types update, that's where I'm lost. I tried a few things, but they didn't look right. so I figured I'd leave the joy of types to yourself, as you are the expert in types resolution, to figure out how to update the signature properly...

@RebeccaStevens
